### PR TITLE
Consistent backend interface with documentation

### PR DESCRIPTION
**Pull request by benlangfeld**
Special attention when reviewing should be paid to the methods that were made private or removed on each backend implementation. When reviewing, taking each commit one at a time may be helpful.

### CODE REVIEW

**Comment by SamSaffron** [comment]
Do we need the giant indenting change? also the huge change in formatting makes it super hard to pick up on non indenting changes.

**Comment by benlangfeld** [comment]
@SamSaffron The indenting change is all in the first commit. If you review the commits one by one, it's easy to follow what's going on. The indenting is in order to use the style of module/class declaration (nested) that the Ruby community understands as preferred.

File: lib/message_bus/backends/memory.rb [hunk 1]
```diff
@@ -190,39 +190,36 @@ def initialize(config = {}, max_backlog_size = 1000)
...
+      # @param [String] channel the name of the channel to which the message should be published
+      # @param [JSON] data some data to publish to the channel. Must be an object that can be encoded as JSON
+      # @param [Hash] opts
+      # @option opts [Boolean] :queue_in_memory NOT SUPPORTED
+      # @option opts [Integer] :max_backlog_age (`self.max_backlog_age`) the longest amount of time a message may live in a backlog before beging removed, in seconds
+      # @option opts [Integer] :max_backlog_size (`self.max_backlog_size`) the largest permitted size (number of messages) for the channel backlog; beyond this capacity, old messages will be dropped
+      #
+      # @return [Integer] the channel-specific ID the message was given
       def publish(channel, data, opts = nil)
-        client = self.client
```
Comment (SamSaffron) [inline lib/message_bus/backends/memory.rb]: this was there to erase method calls

File: lib/message_bus/backends/postgres.rb [hunk 2]
```diff
@@ -237,41 +238,34 @@ def initialize(config = {}, max_backlog_size = 1000)
...
+      # @param [Hash] opts
+      # @option opts [Boolean] :queue_in_memory NOT SUPPORTED
+      # @option opts [Integer] :max_backlog_age (`self.max_backlog_age`) the longest amount of time a message may live in a backlog before beging removed, in seconds
+      # @option opts [Integer] :max_backlog_size (`self.max_backlog_size`) the largest permitted size (number of messages) for the channel backlog; beyond this capacity, old messages will be dropped
+      #
+      # @return [Integer] the channel-specific ID the message was given
       def publish(channel, data, opts = nil)
         # TODO in memory queue?
 
-        client = self.client
```
Comment (SamSaffron) [inline lib/message_bus/backends/postgres.rb]: same

File: lib/message_bus/backends/base.rb [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+  module Backends
+    # The heart of the message bus, it acts as two things:
+    #
+    # 1. A channel multiplexer
+    # 2. Backlog storage per-multiplexed channel.
+    #
+    # Message ids are all sequencially increasing integers starting at 0.
+    #
+    # @abstract
+    class Base
```
Comment (SamSaffron) [inline lib/message_bus/backends/base.rb]: 👍  on having a new base interface

File: lib/message_bus/backends/memory.rb (hunk 1 above)
Comment (benlangfeld) [reply lib/message_bus/backends/memory.rb]: How so? I don't see how it's different.

**Comment by SamSaffron** [comment]
👍  on all 3 of Jeremy's points. 

I also worry about exact duplicate documentation it will go out of sync, I think you can have links in documentation for yard so just use that . 

Once we get these points addressed I am happy to merge this in.

**Comment by benlangfeld** [comment]
1. Done. I opted for raising a custom exception because otherwise I don't have a good way to test compliance with the interface.
2. Done (see https://github.com/lsegal/yard/issues/720 for details).
3. Done.

Will merge this then.

**Pull request merged**